---
title: "Single cell RNA-seq clustering - Basic"
output: html_notebook
---


```{r init-workspace, message=FALSE, warning=FALSE, include=FALSE}
if (! require(Rtsne)) {
  install.packages("Rtsne")
  require(Rtsne)
}

CountMarkerGenes <- function(expr.dat, clusters, 
                             expr.thresh = 1, cl1.prop = 0.5, cl2.prop = 0.1) {
  
  cl.sizes <- table(clusters)
  cl.props <- t(apply(expr.dat, 1, function(x) {
    tapply(x, clusters, function(y) sum(y > expr.thresh) / length(y))
    }))
  
  min.dex.df <- data.frame()
  if (sum(table(clusters) > 1) > 1) {
    for (i in 1:(ncol(cl.props) - 1)) {
      for (j in (i + 1):ncol(cl.props)) {
        dex.cnt1 <- sum(cl.props[, i] > cl1.prop & cl.props[, j] < cl2.prop)
        dex.cnt2 <- sum(cl.props[, j] > cl1.prop & cl.props[, i] < cl2.prop)
        min.dex.cnt <- min(dex.cnt1, dex.cnt2)
        min.size <- min(cl.sizes[i], cl.sizes[j])
        cl1 <- as.numeric(colnames(cl.props)[i])
        cl2 <- as.numeric(colnames(cl.props)[j])
        min.dex1 <- data.frame(cl1, cl2, min.size, min.dex.cnt)
        min.dex.df <- rbind(min.dex.df, min.dex1)
                            
      }
    }
  }

  min.dex.df <- min.dex.df[order(min.dex.df$min.dex.cnt), ]
  return(min.dex.df)
}

```

## Load data
#### Data downloaded from 
```{r load-data, echo = FALSE}
expr <- read.csv(file = "data/fpkm_table.csv", row.names = 1)
genes <- read.csv(file = "data/rows-genes.csv", row.names = 1)
meta <- read.csv(file = "data/columns-cells.csv")

```

## Subset data
```{r subset-data}
# Keep cells with at least N genes detected
min.genes <- 1000
keep.cells <- which(apply(expr, 2, function(x) sum(x > 0)) > min.genes)

# Select random subset of cells to speed computation
set.seed(123)
keep.cells <- sample(keep.cells, size = 100, replace = FALSE)

# Keep annotated, variable genes
var.genes <- apply(expr[, keep.cells], 1, sd) > 0
keep.genes <- which(! duplicated(genes$gene_symbol) & 
                      ! grepl("^Gm", genes$gene_symbol) &
                      ! grepl("^LOC", genes$gene_symbol) &
                      var.genes)

# Subset data
genes.subset <- droplevels(genes[keep.genes, ])
expr.subset <- log2(expr[keep.genes, keep.cells] + 1)
meta.subset <- droplevels(meta[keep.cells, ])
rownames(expr.subset) <- genes.subset$gene_symbol
rownames(meta.subset) <- colnames(expr.subset)

```

## Select variable genes
```{r select-variable-genes}
gene.sd <- apply(expr.subset, 1, sd)
var.thresh <- quantile(gene.sd, 0.95)
hist(gene.sd, breaks = 20, xlab = "Expression standard deviation", 
     main = "Variable genes")
abline(v = var.thresh, col = "blue")
top.var.genes <- which(gene.sd > var.thresh)

```

## Reduce dimensionality
```{r reduce-dim}
expr.scaled <- scale(expr.subset)
pca1 <- prcomp(expr.scaled[top.var.genes, ])

```

## Select principal components
```{r select-pcs}
plot(pca1, type = "l")
sig.pcs <- 1:4
expr.pcs <- pca1$rotation[, sig.pcs]

```

## Cluster cells
```{r find-clusters}
km.list <- list()
within.ss <- NULL
for (i in 1:10) {
  km1 <- kmeans(expr.pcs, centers = i, iter.max = 100, nstart = 100)
  km.list[[i]] <- km1
  within.ss[i] <- km1$tot.withinss
}

```

## Select cluster number
```{r select-clusters}
plot(1:10, within.ss, type="b",
     xlab="Number of Clusters", ylab="Within groups sum of squares")

for (i in 2:10) {
  cl.lab <- km.list[[i]]$cluster
  pcs.toplot <- 1:min(4, ncol(expr.pcs))
  pairs(expr.pcs[, pcs.toplot], col = cl.lab, main = paste(i, "clusters"))
}

n.clus <- 10
meta.subset$cluster <- km.list[[n.clus]]$cluster
meta.subset$cluster_curated <- meta.subset$cluster

```

## Curate clusters
```{r curate-clusters}
# Cell type criteria
p.expr.thresh <- 1  # Expression noise floor
target.prop <- 0.5  # Define marker gene - minimum detection in target cluster
other.prop <- 0.1  # Define marker gene - maximum detection in other clusters
min.markers <- 3  # Minimum number of marker genes to identify distinct cell type
min.cl.size <- 3  # Minimum number of cells in a cluster

cl.marker.cnt <- CountMarkerGenes(expr.subset, meta.subset$cluster_curated,
                                  expr.thresh = p.expr.thresh,
                                  cl1.prop = target.prop, cl2.prop = other.prop)
print(head(cl.marker.cnt))

# Merge clusters
cl.tomerge <- which(cl.marker.cnt$min.size < min.cl.size |
                      cl.marker.cnt$min.dex.cnt < min.markers)
for (idx1 in cl.tomerge) {
  cl.from <- cl.marker.cnt$cl2[idx1]
  cl.to <- cl.marker.cnt$cl1[idx1]
  meta.subset$cluster_curated[meta.subset$cluster_curated == cl.from] <- cl.to
}

```


## Visualize clustering
```{r viz-clusters, fig.height=8}
par(mfrow = c(2, 2))
plot(expr.pcs[, 1:2], col = meta.subset$cluster, 
      main = paste(length(unique(meta.subset$cluster)), "original clusters"))
plot(expr.pcs[, 1:2], col = meta.subset$cluster_curated, 
      main = paste(length(unique(meta.subset$cluster_curated)), "curated clusters"))

for (clus1 in c("cluster", "cluster_curated")) {
  cl.expr <- t(apply(expr.subset[top.var.genes, ], 1, 
                     function(x) tapply(x, meta.subset[, clus1], mean)))
  cl.dist <- as.dist((1 - cor(cl.expr)) / 2)
  cl.dend <- as.dendrogram(hclust(cl.dist))
  plot(cl.dend)
}

tsne1 <- Rtsne(expr.pcs, perplexity = 30)$Y
plot(tsne1, col = meta.subset$cluster_curated, 
      main = paste(length(unique(meta.subset$cluster_curated)), "curated clusters"))


```
