---
title: "Single cell RNA-seq clustering - Basic"
output: html_notebook
---


```{r init-workspace, message=FALSE, warning=FALSE, include=FALSE}
# Install required packages
if (! require(Rtsne)) {
  install.packages("Rtsne")
  require(Rtsne)
}

if (! require(RColorBrewer)) {
  install.packages("RColorBrewer")
  require(RColorBrewer)
}
palette(brewer.pal(12, "Set3"))


# Define functions
CalcBrokenStick <- function(x) {
  m <- 0
  out <- matrix(NA, ncol = 3, nrow = length(x))
  colnames(out) <- c("pc", "var_pct", "bstick_thresh")
  for (i in 1:length(x)) {
    for (k in i:length(x)) {
      m <- m + ((1 / length(x)) * (1 / k))
    }
    out[i, ] <- c(i, (x[i] / sum(x)) * 100, m * 100)
    m <- 0
  }
  pc.var.order <- order(out[, "var_pct"], decreasing = TRUE)
  out <- out[pc.var.order, ]
  out[, "bstick_thresh"] <- sort(out[, "bstick_thresh"], decreasing = TRUE)
  return(out)
}

CountMarkerGenes <- function(expr.dat, clusters, 
                             expr.thresh = 1, cl1.prop = 0.5, cl2.prop = 0.1) {
  
  cl.sizes <- table(clusters)
  cl.props <- t(apply(expr.dat, 1, function(x) {
    tapply(x, clusters, function(y) sum(y > expr.thresh) / length(y))
    }))
  
  min.dex.df <- data.frame()
  if (sum(table(clusters) > 1) > 1) {
    for (i in 1:(ncol(cl.props) - 1)) {
      for (j in (i + 1):ncol(cl.props)) {
        dex.cnt1 <- sum(cl.props[, i] > cl1.prop & cl.props[, j] < cl2.prop)
        dex.cnt2 <- sum(cl.props[, j] > cl1.prop & cl.props[, i] < cl2.prop)
        min.dex.cnt <- min(dex.cnt1, dex.cnt2)
        min.size <- min(cl.sizes[i], cl.sizes[j])
        cl1 <- colnames(cl.props)[i]
        cl2 <- colnames(cl.props)[j]
        min.dex1 <- data.frame(cl1, cl2, min.size, min.dex.cnt)
        min.dex.df <- rbind(min.dex.df, min.dex1)
                            
      }
    }
    min.dex.df <- min.dex.df[order(min.dex.df$min.dex.cnt), ]
  }

  return(min.dex.df)
}

```

## Load data
300 cells from mouse dorsal lateral geniculate nucleus of the thalamus. 
The full data set (1772 cells) is available for download from the 
Allen Institute Cell Types Database: http://celltypes.brain-map.org/download.
```{r load-data, echo = FALSE}
expr <- read.csv(file = "data/fpkm_table.csv", row.names = 1)
genes <- read.csv(file = "data/rows-genes.csv")
meta <- read.csv(file = "data/columns-cells.csv")

```

## Subset data
```{r subset-data}
# Keep cells with at least N genes detected
min.genes <- 1000
keep.cells <- which(apply(expr, 2, function(x) sum(x > 0)) > min.genes)

# Keep annotated, variable genes
var.genes <- apply(expr[, keep.cells], 1, sd) > 0
keep.genes <- which(! duplicated(genes$gene_symbol) & 
                      ! grepl("^Gm", genes$gene_symbol) &
                      ! grepl("^LOC", genes$gene_symbol) &
                      var.genes)

# Subset data
genes.subset <- droplevels(genes[keep.genes, ])
expr.subset <- log2(expr[keep.genes, keep.cells] + 1)
meta.subset <- droplevels(meta[keep.cells, ])
rownames(expr.subset) <- genes.subset$gene_symbol
rownames(meta.subset) <- colnames(expr.subset)

```

## Select variable genes
```{r select-variable-genes}
gene.mean <- apply(expr.subset, 1, mean)
gene.sd <- apply(expr.subset, 1, sd)
sd.thresh <- quantile(gene.sd, 0.95)

plot(gene.mean, gene.sd, cex = 0.5, 
     xlab = "Average expression - log2(FPKM + 1)", 
     ylab = "Expression variation (sd)", 
     main = "Variable genes")
abline(h = sd.thresh, col = "blue")
top.var.genes <- which(gene.sd > sd.thresh)

# plot(gene.mean^2, (gene.sd / gene.mean)^2, log = "xy")

```

## Reduce dimensionality
```{r reduce-dim}
expr.scaled <- scale(expr.subset)
pca1 <- prcomp(expr.scaled[top.var.genes, ])

```

## Select principal components
```{r select-pcs}
# Find PCs with more variance explained than broken stick distribution
brstick1 <- CalcBrokenStick(pca1$sdev^2)

sig.pcs <- NULL
for (i in 1:10) {
  var.pct <- brstick1[i, "var_pct"]
  var.thresh <- brstick1[i, "bstick_thresh"]
  if (var.pct > var.thresh) {
    sig.pcs <- c(sig.pcs, brstick1[i, "pc"])
  } else {
    break
  }
}

plot(pca1, type = "l", main = "Explained variance of PCs")
abline(v = max(sig.pcs), col = "blue")
expr.pcs <- pca1$rotation[, sig.pcs]

```

## Cluster cells
```{r find-clusters-kmeans}
# K-means clustering
km.list <- list()
within.ss <- NULL
for (i in 1:20) {
  km1 <- kmeans(expr.pcs, centers = i, iter.max = 100, nstart = 100)
  km.list[[i]] <- km1
  within.ss[i] <- km1$tot.withinss
}

```

```{r find-clusters-graph, eval = FALSE}
# Graph-based clustering (Jaccard Louvain)
require(Rphenograph)

nn.num <- 15  # Number of nearest cells to compare in building graph
rpheno <- Rphenograph(expr.pcs, k = nn.num)

```


## Select cluster number
```{r select-clusters}
plot(1:20, within.ss, type="b",
     xlab="Number of Clusters", ylab="Within groups sum of squares")

tsne1 <- Rtsne(expr.pcs, perplexity = 30)$Y

par(mfrow = c(3, 4), mar = c(1, 2, 4, 2))
for (i in 2:13) {
  cl.lab <- km.list[[i]]$cluster
  plot(tsne1, col = cl.lab, main = paste(i, "clusters"))
  # PCA plots
  # pcs.toplot <- 1:min(4, ncol(expr.pcs))
  # pairs(expr.pcs[, pcs.toplot], col = cl.lab, main = paste(i, "clusters"))
}

n.clus <- 10
# meta.subset$cluster <- km.list[[n.clus]]$cluster
meta.subset$cluster <- membership(rpheno[[2]])
meta.subset$cluster_curated <- meta.subset$cluster

```

## Curate clusters
```{r curate-clusters}
# Cell type criteria
p.expr.thresh <- 1  # Marker gene: Min expression for detection above noise
target.prop <- 0.5  # Marker gene: Min detection (proportion of cells) in target cluster
other.prop <- 0.1  # Marker gene: Max detection in other clusters
min.markers <- 3  # Min number of marker genes
min.cl.size <- 3  # Min number of cells

cl.marker.cnt <- CountMarkerGenes(expr.subset, meta.subset$cluster_curated,
                                  expr.thresh = p.expr.thresh,
                                  cl1.prop = target.prop, cl2.prop = other.prop)
print(head(cl.marker.cnt))

# Merge similar clusters
cl.tomerge <- which(cl.marker.cnt$min.size < min.cl.size |
                      cl.marker.cnt$min.dex.cnt < min.markers)
for (idx1 in cl.tomerge) {
  cl.from <- cl.marker.cnt$cl2[idx1]
  cl.to <- cl.marker.cnt$cl1[idx1]
  meta.subset$cluster_curated[meta.subset$cluster_curated == cl.from] <- cl.to
}

```


## Visualize clusters
```{r viz-clusters}
par(mfrow = c(2, 3))

for (clus1 in c("cluster", "cluster_curated")) {
  plot(expr.pcs[, 1:2], col = meta.subset[, clus1],
      main = paste("PCA -", length(unique(meta.subset[, clus1])), clus1))
  plot(tsne1, col = meta.subset[, clus1], 
      main = paste("tSNE -", length(unique(meta.subset[, clus1])), clus1))

  cl.expr <- t(apply(expr.subset[top.var.genes, ], 1, 
                     function(x) tapply(x, meta.subset[, clus1], mean)))
  cl.dist <- as.dist((1 - cor(cl.expr)) / 2)
  cl.dend <- as.dendrogram(hclust(cl.dist))
  
  # Reorder dendrogram to match cluster position along first tSNE coordinate
  cl.coord <- tapply(tsne1[, 1], meta.subset[, clus1], mean)
  cl.dend <- reorder(cl.dend, cl.coord, agglo.FUN = "mean")
  plot(cl.dend,
             main = paste(length(unique(meta.subset[, clus1])), clus1))
}

```



## Further analyses
1. Try different clustering parameters
    * Variable gene selection
    * Number of principal components
2. Split clusters further by clustering subsets of cells
3. Plot expression of cluster markers
4. Download and cluster complete mouse dorsal LGN data set (1,772 cells)
    * http://celltypes.brain-map.org/api/v2/well_known_file_download/502999251


## Resources
1. RNA-seq data sets
    * Allen Institute Cell Types Database http://celltypes.brain-map.org/download
    * Single Cell Portal (Broad Institute) https://portals.broadinstitute.org/single_cell
    * SCAP-T https://www.scap-t.org/content/data-portal
    * NCBI GEO DataSets https://www.ncbi.nlm.nih.gov/gds
2. Analysis tools
    * Cell sampling http://satijalab.org/howmanycells
    * BASiCS https://github.com/catavallejos/BASiCS
    * RUVSeq http://bioconductor.org/packages/release/bioc/html/RUVSeq.html
    * DESeq2 https://bioconductor.org/packages/release/bioc/html/DESeq2.html
    * scde http://hms-dbmi.github.io/scde/
    * WGCNA https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/
    * tSNE https://lvdmaaten.github.io/tsne/
    * ToppGene GO enrichment https://toppgene.cchmc.org/enrichment.jsp
3. Clustering tools
    * DBSCAN https://cran.r-project.org/web/packages/dbscan/
    * Pagoda https://github.com/hms-dbmi/pagoda2
    * Seurat http://satijalab.org/seurat/
    * BackSpin https://github.com/linnarsson-lab/BackSPIN
    * PhenoGraph https://www.c2b2.columbia.edu/danapeerlab/html/phenograph.html
    * SIMLR https://github.com/BatzoglouLabSU/SIMLR
