---
title: "Single cell RNA-seq clustering"
output: html_notebook
---


Load required packages and functions.
```{r init-workspace, warning=FALSE}
#### Install packages ####
if (! require(Rtsne)) {
  install.packages("Rtsne")
  require(Rtsne)
}

if (! require(scde)) {
  install.packages("scde")
  require(scde)
}

if (! require(Rphenograph)) {
  if (! require(devtools)) {
    install.packages("devtools")
  }
  devtools::install_github("JinmiaoChenLab/Rphenograph")
}


#### Load functions ####
source(file = "fCalcBrokenStick.R")
source(file = "fCountMarkerGenes.R")
source(file = "fFitNoiseModels.R")


#### Set color palette ####
pal1 <- c("#d77a7f", "#8eda48", "#7340cd", "#d6c847", "#ce4cc5", "#64db8e", 
          "#432876", "#509140", "#7171cd", "#d1863a", "#79acd9", "#d24530", 
          "#6dc7b7", "#d23e70", "#c6d394", "#8d3870", "#827f38", "#cd90cb", 
          "#3a4e32", "#c9c5c6", "#3e263b", "#ae8875", "#556983", "#753627")
palette(pal1)
```

## Load data
This data set includes 300 nuclei, 250 NeuN+ (neuronal) and 50 NeuN- 
(non-neuronal), dissected from Layer 5 of human neocortex (middle temporal gyrus). 
The full data set (15,928 nuclei) can be downloaded from the Allen Institute 
Cell Types Database: http://celltypes.brain-map.org/download.
```{r load-data, echo = FALSE}
expr <- read.csv(file = "../data/counts_table.csv", row.names = 1)
meta <- read.csv(file = "../data/columns-nuclei.csv")

```

## Subset data
```{r subset-data}
# Keep cells with at least M genes detected
min.genes <- 1000
keep.samp <- which(apply(expr, 2, function(x) sum(x > 0)) > min.genes)

# Keep genes with some variability
keep.genes <- which(apply(expr[, keep.samp], 1, sd) > 0)

# Subset and log-transform counts per million
expr.cnt <- expr[keep.genes, keep.samp]
expr.cpm <- round(sweep(expr.cnt, 2, colSums(expr.cnt), "/") * 1e6, 0)
expr.cpm <- apply(expr.cpm, 2, function(x) { storage.mode(x) <- 'integer'; x })
expr.log <- log2(expr.cpm + 1)
meta.subset <- droplevels(meta[keep.samp, ])
rownames(meta.subset) <- colnames(expr.cnt)

```

## Select variable genes
```{r select-variable-genes, fig.height = 4, fig.width = 8}
expr.mean <- apply(expr.cpm, 1, mean)
expr.mean.bin <- cut(expr.mean, breaks = quantile(expr.mean, seq(0, 1, 1/20)),
                     include.lowest = TRUE)
expr.cv <- apply(expr.cpm, 1, function(x) sd(x) / mean(x))
expr.cv.z.list <- tapply(expr.cv, expr.mean.bin, scale)
expr.cv.z <- do.call("rbind", expr.cv.z.list)
expr.cv.z <- expr.cv.z[match(names(expr.mean), rownames(expr.cv.z)), 1]
expr.cv.z[expr.cv.z > 10] <- 10

# Select z-score threshold for variable genes
z.thresh <- 2
top.var.genes <- which(expr.cv.z > z.thresh)

par(mfrow = c(1, 2))
plot(expr.mean, expr.cv, cex = 0.5, log = "xy", 
     xlab = "Average expression", 
     ylab = "Coefficient of variation (CV)", 
     main = "Top variable genes")
points(expr.mean[top.var.genes], expr.cv[top.var.genes], col = "blue")

plot(expr.mean, expr.cv.z, cex = 0.5, log = "x", 
     xlab = "Average expression", 
     ylab = "CV (z-score)", 
     main = "Top variable genes")
points(expr.mean[top.var.genes], expr.cv.z[top.var.genes], col = "blue")

```

## Fit noise models
```{r fit-noise-models}
noise.models <- FitNoiseModels(expr.cpm)


```


## Reduce dimensionality
```{r reduce-dim}
expr.scaled <- scale(expr.cpm)
pca1 <- prcomp(expr.scaled[top.var.genes, ])

```

## Select principal components
```{r select-pcs}
# Find PCs with more variance explained than broken stick distribution
brstick1 <- CalcBrokenStick(pca1$sdev^2)

sig.pcs <- NULL
for (i in 1:10) {
  var.pct <- brstick1[i, "var_pct"]
  var.thresh <- brstick1[i, "bstick_thresh"]
  if (var.pct > var.thresh) {
    sig.pcs <- c(sig.pcs, brstick1[i, "pc"])
  } else {
    break
  }
}

# sig.pcs <- 1:2  # Manually specify PCs
plot(pca1, type = "l", main = "Explained variance of PCs")
abline(v = max(sig.pcs), col = "blue")
expr.pcs <- pca1$rotation[, sig.pcs]

```

## Cluster cells
```{r find-clusters-kmeans}
# K-means clustering
km.list <- list()
within.ss <- NULL
for (i in 1:20) {
  km1 <- kmeans(expr.pcs, centers = i, iter.max = 100, nstart = 100)
  km.list[[i]] <- km1
  within.ss[i] <- km1$tot.withinss
}

plot(1:20, within.ss, type="b",
     xlab="Number of Clusters", ylab="Within groups sum of squares")

# tSNE dimensionality reduction for cluster visualization
tsne1 <- Rtsne(expr.pcs, perplexity = 20)$Y

par(mfrow = c(3, 4), mar = c(1, 1, 4, 1))
for (i in 2:13) {
  cl.lab <- km.list[[i]]$cluster
  plot(tsne1, col = cl.lab, main = paste(i, "clusters"))
  # plot(expr.pcs[, 1:2], col = cl.lab, main = paste(i, "clusters"))
}

```

```{r find-clusters-graph}
# Graph-based clustering (Jaccard/Louvain)
nn.num <- 15  # Number of nearest cells to compare in building graph
rpheno <- Rphenograph(expr.pcs, k = nn.num)

plot.lab <- paste(length(unique(membership(rpheno[[2]]))), "clusters; ",
                  round(modularity(rpheno[[2]]), 2), "modularity")
plot(rpheno[[1]], vertex.size = 10, vertex.label = NA, 
     vertex.color = pal1[membership(rpheno[[2]])], main = plot.lab)

```


## Select cluster number
```{r select-clusters}
# Select kmeans clusters
n.clus <- 11  # Number of clusters
# cell.cl <- km.list[[n.clus]]$cluster

# Select graph-based clusters
cell.cl <- membership(rpheno[[2]])

# Select clusters identified based on whole data set (1772 cells)
# cell.cl <- meta.subset$subclass

# Store cluster membership
meta.subset$cluster <- as.character(cell.cl)
meta.subset$cluster_curated <- as.character(cell.cl)

```

## Curate clusters
```{r curate-clusters}
# Cell type criteria
p.expr.thresh <- 1  # Marker gene: Min expression for detection above noise
target.prop <- 0.5  # Marker gene: Min detection (proportion of cells) in target cluster
other.prop <- 0.1  # Marker gene: Max detection in other clusters
min.markers <- 3  # Min number of marker genes
min.cl.size <- 3  # Min number of cells

# Iteratively merge clusters based on above criteria
merge.cl <- TRUE
while (merge.cl) {
  cl.marker.cnt <- CountMarkerGenes(expr.cpm, meta.subset$cluster_curated,
                                    expr.thresh = p.expr.thresh,
                                    cl1.prop = target.prop, cl2.prop = other.prop)
  
  cl.tomerge <- which(cl.marker.cnt$min.size < min.cl.size |
                        cl.marker.cnt$min.dex.cnt < min.markers)
  
  if (length(cl.tomerge) > 0) {
    for (idx1 in cl.tomerge) {
      cl.from <- cl.marker.cnt$cl2[idx1]
      cl.to <- cl.marker.cnt$cl1[idx1]
      meta.subset$cluster_curated[meta.subset$cluster_curated == cl.from] <- cl.to
    }
  } else {
    merge.cl <- FALSE
  }
}

```


## Visualize final clusters
```{r viz-clusters}
par(mfrow = c(2, 3))

for (clus1 in c("cluster", "cluster_curated")) {
  plot.color <- pal1[as.factor(meta.subset[, clus1])]
  plot(expr.pcs[, 1:2], col = plot.color,
      main = paste("PCA -", length(unique(meta.subset[, clus1])), clus1))
  plot(tsne1, col = plot.color, xlab = "tSNE1", ylab = "tSNE2", 
      main = paste("tSNE -", length(unique(meta.subset[, clus1])), clus1))

  cl.expr <- t(apply(expr.cpm[top.var.genes, ], 1, 
                     function(x) tapply(x, meta.subset[, clus1], mean)))
  cl.dist <- as.dist((1 - cor(cl.expr)) / 2)
  cl.dend <- as.dendrogram(hclust(cl.dist))
  
  # Reorder dendrogram to match cluster position along first tSNE coordinate
  cl.coord <- tapply(tsne1[, 1], meta.subset[, clus1], mean)
  cl.dend <- reorder(cl.dend, cl.coord, agglo.FUN = "mean")
  plot(cl.dend, main = paste(length(unique(meta.subset[, clus1])), clus1))
}

```


## Further analyses
1. Cluster with different parameters
    * Adjust threshold for variable genes
    * Select more/fewer principal components
    * Change criteria for merging cell types
2. Rerun clustering on cells in one cluster to try to further divide types
3. Download and cluster complete mouse dorsal LGN data set (1,772 cells)
    * http://celltypes.brain-map.org/api/v2/well_known_file_download/502999251
4. Identify and plot expression of cluster markers

## Resources
1. RNA-seq data sets
    * Allen Institute Cell Types Database http://celltypes.brain-map.org/download
    * Single Cell Portal (Broad Institute) https://portals.broadinstitute.org/single_cell
    * SCAP-T https://www.scap-t.org/content/data-portal
    * NCBI GEO DataSets https://www.ncbi.nlm.nih.gov/gds
2. Analysis tools
    * Cell sampling http://satijalab.org/howmanycells
    * BASiCS https://github.com/catavallejos/BASiCS
    * RUVSeq http://bioconductor.org/packages/release/bioc/html/RUVSeq.html
    * DESeq2 https://bioconductor.org/packages/release/bioc/html/DESeq2.html
    * scde http://hms-dbmi.github.io/scde/
    * WGCNA https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/
    * tSNE https://lvdmaaten.github.io/tsne/
    * ToppGene GO enrichment https://toppgene.cchmc.org/enrichment.jsp
3. Clustering tools
    * DBSCAN https://cran.r-project.org/web/packages/dbscan/
    * Pagoda https://github.com/hms-dbmi/pagoda2
    * Seurat http://satijalab.org/seurat/
    * BackSpin https://github.com/linnarsson-lab/BackSPIN
    * PhenoGraph https://www.c2b2.columbia.edu/danapeerlab/html/phenograph.html
    * SIMLR https://github.com/BatzoglouLabSU/SIMLR
